### PR TITLE
enable cache & add detach when destroy computation

### PR DESCRIPTION
**Pull request by yanwei-gr**
Run a model will cache the compiled executable file to the target path. This will be used to accelerate the start up process when open a new session to run the same model next time.

### CODE REVIEW

File: ODLA/platforms/odla_popart/odla_compute.cc [hunk 1]
```diff
@@ -51,6 +52,7 @@
 #endif
 
 thread_local odla_computation g_comp;
+thread_local volatile bool comp_initialized = false;
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/odla_compute.cc:55]: put this field into odla_computation_ structure?

File: ODLA/platforms/odla_popart/odla_compute.cc [hunk 2]
```diff
@@ -168,26 +164,65 @@ odla_status odla_DestroyContext(odla_context ctx) {
...
+    comp->session.reset();
+  }
   // g_comp.reset();
   return ODLA_SUCCESS;
 }
 
 odla_status odla_ExecuteComputation(odla_computation comp, odla_context context,
                                     odla_compute_mode mode,
                                     odla_device device) {
+  if (!comp_initialized) {
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/odla_compute.cc:179]: it will make the first execution very slow. If we do initialization "odla_CreateContext" like before, we can avoid this problem

File: ODLA/platforms/odla_popart/odla_compute.cc [hunk 3]
```diff
@@ -147,28 +167,35 @@ odla_status odla_CreateContext(odla_context* context) {
...
   return ODLA_SUCCESS;
 }
 
 odla_status odla_DestroyComputation(odla_computation comp) {
-  // g_comp.reset();
+  if (comp->session != nullptr) {
+    comp->session->getDevice().getDeviceInfo()->detach();
+    comp->session.reset();
+  }
+  if (g_comp = nullptr) {
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/odla_compute.cc:196]: something like
```
if (g_comp == comp) { g_comp == nullptr;}
delete comp;
```
?

File: ODLA/platforms/odla_popart/odla_popart.h [hunk 4]
```diff
@@ -64,8 +64,7 @@ struct _odla_value {
 
 struct _odla_computation {
   std::unique_ptr<popart::Builder> builder;
-  std::map<popart::TensorId, std::unique_ptr<popart::IArray>> inputs;
-  std::map<popart::TensorId, std::unique_ptr<popart::IArray>> outputs;
+  std::unique_ptr<popart::InferenceSession> session;
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/odla_popart.h:67]: can we have `odla_context` in `odla_computation` and `session` is saved in `_odla_context`?
like 
```
struct _odla_computation {
   odla_context ctx;
   ...
};
```


Then, in `odla_DestroyContext(ctx)`, we can do  
```
// reset ctx->session
ctx->comp->ctx = nullptr; 
```

In `odla_DestroyComputation(comp)`, we do 
```
if (comp->ctx) { odla_DestroyContext(comp->ctx); } 
delete (comp);
```

File: utils/docker/start_docker_cpu.sh [hunk 5]
```diff
@@ -25,7 +25,7 @@ GROUPID=`id -g`
 OLD_ID=`docker ps -aq -f name=$CONTAINER_NAME -f status=running`
 
 if [ -z "$OLD_ID" ]; then
-  ID=`docker run $docker_run_flag --privileged --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -t -d --name $CONTAINER_NAME -v $MOUNT_DIR:/host  --tmpfs /tmp:exec --rm $IMAGE `
+  ID=`docker run $docker_run_flag --privileged --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -t -d --name $CONTAINER_NAME -v $MOUNT_DIR:/host -v /home:/home -v /data:/data --tmpfs /tmp:exec --rm $IMAGE `
```
Comment (weimingzha0) [inline utils/docker/start_docker_cpu.sh:28]: revert this change?

File: ODLA/platforms/odla_popart/odla_compute.cc (hunk 3 above)
Comment (yanwei-gr) [reply ODLA/platforms/odla_popart/odla_compute.cc]: It seems that set `g_comp=nullptr` and `delete comp` or `comp = nullptr` together will cause a segment error.

File: utils/docker/start_docker_cpu.sh (hunk 5 above)
Comment (yanwei-gr) [reply utils/docker/start_docker_cpu.sh]: reverted

File: ODLA/platforms/odla_popart/odla_popart.h (hunk 4 above)
Comment (yanwei-gr) [reply ODLA/platforms/odla_popart/odla_popart.h:67]: I didn't change this based on our discussion in the morning. If I understand our conclusion correctly. I keep `session` in computation and do the destruction as before...Please correct me if I'm wrong.

File: ODLA/platforms/odla_popart/odla_compute.cc [hunk 6]
```diff
@@ -147,28 +167,37 @@ odla_status odla_CreateContext(odla_context* context) {
...
+  }
   return ODLA_SUCCESS;
 }
 
 odla_status odla_DestroyComputation(odla_computation comp) {
-  // g_comp.reset();
+  if (comp->session != nullptr) {
+    comp->session->getDevice().getDeviceInfo()->detach();
+    comp->session.reset();
+  }
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/odla_compute.cc:195]: delete comp ?

File: ODLA/platforms/odla_popart/odla_compute.cc [hunk 7]
```diff
@@ -147,28 +167,37 @@ odla_status odla_CreateContext(odla_context* context) {
 
   // Create InferenceSession
   auto proto = g_comp->builder->getModelProto();
-  auto session = popart::InferenceSession::createFromOnnxModel(
+  g_comp->session = popart::InferenceSession::createFromOnnxModel(
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/odla_compute.cc:170]: can we check if g_comp->session has already been created or not?
we will call 
 "odla_CreatContext() -- odla_DestroyContext() " multiple times.

The "inferenceSession" has the same lifecycle as computation, is that correct?
we just delay the creation of popart:inferenceSession to the first odla_CreateContext()

**Review by weimingzha0 — APPROVED** [review]

File: ODLA/platforms/odla_popart/odla_compute.cc [hunk 8]
```diff
@@ -101,6 +108,18 @@ odla_status odla_SetComputationItem(odla_computation comp, odla_item_type type,
...
+                                 odla_executable executable) {
+  return ODLA_SUCCESS;
+}
+
+odla_status odla_LoadExecutable(const odla_char* file_name,
+                                odla_executable* executable,
+                                odla_context* context,
+                                odla_computation* computation) {
+  return ODLA_SUCCESS;
+}
```
Comment (youbeny) [inline ODLA/platforms/odla_popart/odla_compute.cc:121]: odla_StoreExecutable  odla_LoadExecutable  could be removed

**Comment by yanwei-gr** [comment]
re-submit
